man/cmph.1
.TH CMPH "1" "August 2025" "cmph 2.1.0" "User Commands"
.SH NAME
cmph \- minimum perfect hashing tool
.SH SYNOPSIS
.B cmph
[\-vhVgrR] [\-k nkeys] [\-f hash_function] [\-g [\-c value][\-s seed] ] [\-a algorithm]
.br
[\-M memory_in_MB] [\-b BRZ_parameter] [\-d tmp_dir] [\-m file.mph] [\-o c_file]
.br
[\-p c_func_prefix] keysfile
.SH DESCRIPTION
.PP
Command line tool to generate and query minimal perfect hash functions.
.PP
Please refer to http://cmph.sf.net for full documentation.
.TP
\fB\-h\fR
Print an help message
.TP
\fB\-c\fR
This value determines: the number of vertices in the graph for the algorithms
.br
BMZ and CHM.
.br
The number of bits per key required in the FCH algorithm.
.br
The load factor in the CHD_PH algorithm.
.TP
\fB\-a\fR
Algorithm. Valid values are:
.br
chm (default), bmz, bmz8, brz, fch, bdz, bdz_ph, chd, chd_ph
.TP
\fB\-f\fR
hash function (may be used multiple times).
.br
Valid values are: jenkins (default), wyhash, djb2, fnv, sdbm, crc32
.TP
\fB\-V\fR
Print version number and exit
.TP
\fB\-v\fR
Increase verbosity (may be used multiple times)
.TP
\fB\-k\fR
Number of keys
.TP
\fB\-g\fR
Generate a .mph file
.TP
\fB\-C\fR
Generate a C file
.TP
\fB\-r\fR
create an ordering table for the key indices
.TP
\fB\-R\fR
create an ordered keyfile
.TP
\fB\-s\fR
Random seed
.TP
\fB\-m\fR
Minimum perfect hash function .mph file
.TP
\fB\-o\fR
Name of C file. default: stdout (Only for -C)
.TP
\fB\-p\fR
Name of C functions prefix. default: cmph_c_ (Only for -C)
.TP
\fB\-M\fR
Main memory availability (in MB)
.TP
\fB\-d\fR
Temporary directory used in BRZ algorithm
.TP
\fB\-b\fR
Parameter of BRZ algorithm to make the maximal number of keys in a bucket
lower than 256.
.br
For BDZ it is used to determine the size of some precomputed rank
information and its value should be an integer in the range [3,10].
Default is 7. The larger is this value, the more compact are the
resulting functions and the slower are them at evaluation time.
.br

ChangeLog
2005-08-08 18:34  fc_botelho

	* INSTALL, examples/Makefile, examples/Makefile.in,
	examples/.deps/file_adapter_ex2.Po,
	examples/.deps/vector_adapter_ex1.Po, src/brz.c: [no log message]

2005-08-07 22:00  fc_botelho

	* src/: brz.c, brz.h, brz_structs.h, cmph.c, cmph.h, main.c:
	temporary directory passed by command line

2005-08-07 20:22  fc_botelho

	* src/brz.c: stable version of BRZ

2005-08-06 22:09  fc_botelho

	* src/bmz.c: no message

2005-08-06 22:02  fc_botelho

	* src/bmz.c: no message

2005-08-06 21:45  fc_botelho

	* src/brz.c: fastest version of BRZ

2005-08-06 17:20  fc_botelho

	* src/: bmz.c, brz.c, main.c: [no log message]

2005-07-29 16:43  fc_botelho

	* src/brz.c: BRZ algorithm is almost stable

2005-07-29 15:29  fc_botelho

	* src/: bmz.c, brz.c, brz_structs.h, cmph_types.h: BRZ algorithm is
	almost stable

2005-07-29 00:09  fc_botelho

	* src/: brz.c, djb2_hash.c, djb2_hash.h, fnv_hash.c, fnv_hash.h,

src/cmph.c
#include "cmph.h"
#include "cmph_structs.h"
#include "chm.h"
#include "bmz.h"
#include "bmz8.h"
#include "brz.h"
#include "fch.h"
#include "bdz.h"
#include "bdz_ph.h"
#include "chd_ph.h"
#include "chd.h"

#include <stdlib.h>
#include <assert.h>
#include <string.h>
#include <limits.h>
// #define DEBUG
#include "debug.h"

const char *cmph_names[] = {"bmz", "bmz8", "chm", "brz", "fch", "bdz", "bdz_ph", "chd_ph", "chd", NULL };

typedef struct
{
	void *vector;
	cmph_uint32 position; // access position when data is a vector
} cmph_vector_t;



/**
 * Support a vector of struct as the source of keys.
 *
 * E.g. The keys could be the fieldB's in a vector of struct rec where
 * struct rec is defined as:
 * struct rec {
 *    fieldA;
 *    fieldB;
 *    fieldC;
 * }
 */
typedef struct
{
	void *vector;					/* Pointer to the vector of struct */
	cmph_uint32 position; 			/* current position */
	cmph_uint32 struct_size;       	/* The size of the struct */
	cmph_uint32 key_offset;        	/* The byte offset of the key in the struct */
	cmph_uint32 key_len;          	/* The length of the key */
} cmph_struct_vector_t;


static cmph_io_adapter_t *cmph_io_vector_new(void * vector, cmph_uint32 nkeys);
static void cmph_io_vector_destroy(cmph_io_adapter_t * key_source);

static cmph_io_adapter_t *cmph_io_struct_vector_new(void * vector, cmph_uint32 struct_size, cmph_uint32 key_offset, cmph_uint32 key_len, cmph_uint32 nkeys);
static void cmph_io_struct_vector_destroy(cmph_io_adapter_t * key_source);

static int key_nlfile_read(void *data, char **key, cmph_uint32 *keylen)
{
	FILE *fd = (FILE *)data;
	*key = NULL;
	*keylen = 0;
	while(1)
	{
		char buf[BUFSIZ];
		char *c = fgets(buf, BUFSIZ, fd);
		if (c == NULL) return -1;
		if (feof(fd)) return -1;
		*key = (char *)xrealloc(*key, *keylen + strlen(buf) + 1);
		memcpy(*key + *keylen, buf, strlen(buf));
		*keylen += (cmph_uint32)strlen(buf);
		if (buf[strlen(buf) - 1] != '\n') continue;
		break;
	}
	if ((*keylen) && (*key)[*keylen - 1] == '\n')
	{
		(*key)[(*keylen) - 1] = 0;
		--(*keylen);
	}
	return (int)(*keylen);
}

static int key_byte_vector_read(void *data, char **key, cmph_uint32 *keylen)
{
	cmph_vector_t *cmph_vector = (cmph_vector_t *)data;
	cmph_uint8 **keys_vd = (cmph_uint8 **)cmph_vector->vector;
	size_t size;
	memcpy(keylen, keys_vd[cmph_vector->position], sizeof(*keylen));
	size = *keylen;
	*key = (char *)xmalloc(size);
	memcpy(*key, keys_vd[cmph_vector->position] + sizeof(*keylen), size);
	cmph_vector->position = cmph_vector->position + 1;
	return (int)(*keylen);

}

static int key_struct_vector_read(void *data, char **key, cmph_uint32 *keylen)
{
    cmph_struct_vector_t *cmph_struct_vector = (cmph_struct_vector_t *)data;
    char *keys_vd = (char *)cmph_struct_vector->vector;
    cmph_uint64 keys_vd_offset;
    size_t size;
    *keylen = cmph_struct_vector->key_len;
    size = *keylen;
    *key = (char *)xmalloc(size);
    keys_vd_offset = ((cmph_uint64)cmph_struct_vector->position *
                      (cmph_uint64)cmph_struct_vector->struct_size) +
                     (cmph_uint64)cmph_struct_vector->key_offset;
    memcpy(*key, keys_vd + keys_vd_offset, size);
    cmph_struct_vector->position = cmph_struct_vector->position + 1;
    return (int)(*keylen);
}

static int key_vector_read(void *data, char **key, cmph_uint32 *keylen)
{
        cmph_vector_t *cmph_vector = (cmph_vector_t *)data;
        char **keys_vd = (char **)cmph_vector->vector;
	size_t size;
        *keylen = (cmph_uint32)strlen(keys_vd[cmph_vector->position]);
	size = *keylen;
        *key = (char *)xmalloc(size + 1);
        strcpy(*key, keys_vd[cmph_vector->position]);
        cmph_vector->position = cmph_vector->position + 1;
	return (int)(*keylen);
}

static void key_nlfile_dispose(char *key)
{
	free(key);
}

static void key_vector_dispose(char *key)
{
	free(key);
}

static void key_nlfile_rewind(void *data)
{
	FILE *fd = (FILE *)data;
	rewind(fd);
}

static void key_struct_vector_rewind(void *data)
{
	cmph_struct_vector_t *cmph_struct_vector = (cmph_struct_vector_t *)data;
	cmph_struct_vector->position = 0;
}

static void key_vector_rewind(void *data)
{
	cmph_vector_t *cmph_vector = (cmph_vector_t *)data;
	cmph_vector->position = 0;
}

static cmph_uint32 count_nlfile_keys(FILE *fd)
{
    cmph_uint32 count = 0;
    char * ptr;
    rewind(fd);
    while(1)
    {
	char buf[BUFSIZ];
	size_t len;
	ptr = fgets(buf, BUFSIZ, fd);
	if (feof(fd))
	    break;
	if (ferror(fd) || ptr == NULL) {
	    perror("Error reading input file");
	    return 0;
	}
	len = strlen(buf);
	if (len == 0)
	    continue; // skip empty lines
	if (len > BUFSIZ - 1) {
	    fprintf(stderr, "Error: line too long (max %d bytes)\n", BUFSIZ - 1);
	    return 0;
	}
	if (buf[len - 1] != '\n')
	    continue;
	++count;
    }
    rewind(fd);
    return count;
}

cmph_io_adapter_t *cmph_io_nlfile_adapter(FILE * keys_fd)
{
  cmph_io_adapter_t * key_source = (cmph_io_adapter_t *)xmalloc(sizeof(cmph_io_adapter_t));
  assert(key_source);
  key_source->data = (void *)keys_fd;
  key_source->nkeys = count_nlfile_keys(keys_fd);
  key_source->read = key_nlfile_read;
  key_source->dispose = key_nlfile_dispose;
  key_source->rewind = key_nlfile_rewind;
  return key_source;
}

void cmph_io_nlfile_adapter_destroy(cmph_io_adapter_t * key_source)
{
	free(key_source);
}

cmph_io_adapter_t *cmph_io_nlnkfile_adapter(FILE * keys_fd, cmph_uint32 nkeys)
{
  cmph_io_adapter_t * key_source = (cmph_io_adapter_t *)xmalloc(sizeof(cmph_io_adapter_t));
  assert(key_source);
  key_source->data = (void *)keys_fd;
  key_source->nkeys = nkeys;
  key_source->read = key_nlfile_read;
  key_source->dispose = key_nlfile_dispose;
  key_source->rewind = key_nlfile_rewind;
  return key_source;
}

void cmph_io_nlnkfile_adapter_destroy(cmph_io_adapter_t * key_source)
{
	free(key_source);
}


static cmph_io_adapter_t *cmph_io_struct_vector_new(void * vector, cmph_uint32 struct_size, cmph_uint32 key_offset, cmph_uint32 key_len, cmph_uint32 nkeys)
{
	cmph_io_adapter_t * key_source = (cmph_io_adapter_t *)xmalloc(sizeof(cmph_io_adapter_t));
	cmph_struct_vector_t * cmph_struct_vector = (cmph_struct_vector_t *)xmalloc(sizeof(cmph_struct_vector_t));
	assert(key_source);
	assert(cmph_struct_vector);
	cmph_struct_vector->vector = vector;
	cmph_struct_vector->position = 0;
	cmph_struct_vector->struct_size = struct_size;
	cmph_struct_vector->key_offset = key_offset;
	cmph_struct_vector->key_len = key_len;
	key_source->data = (void *)cmph_struct_vector;
	key_source->nkeys = nkeys;
	return key_source;
}

static void cmph_io_struct_vector_destroy(cmph_io_adapter_t * key_source)
{
	cmph_struct_vector_t *cmph_struct_vector = (cmph_struct_vector_t *)key_source->data;
	cmph_struct_vector->vector = NULL;
	free(cmph_struct_vector);
	free(key_source);
}

static cmph_io_adapter_t *cmph_io_vector_new(void * vector, cmph_uint32 nkeys)
{
	cmph_io_adapter_t * key_source = (cmph_io_adapter_t *)xmalloc(sizeof(cmph_io_adapter_t));
	cmph_vector_t * cmph_vector = (cmph_vector_t *)xmalloc(sizeof(cmph_vector_t));
	assert(key_source);
	assert(cmph_vector);
	cmph_vector->vector = vector;
	cmph_vector->position = 0;
	key_source->data = (void *)cmph_vector;
	key_source->nkeys = nkeys;
	return key_source;
}

static void cmph_io_vector_destroy(cmph_io_adapter_t * key_source)
{
	cmph_vector_t *cmph_vector = (cmph_vector_t *)key_source->data;
	cmph_vector->vector = NULL;
	free(cmph_vector);
	free(key_source);
}

cmph_io_adapter_t *cmph_io_byte_vector_adapter(cmph_uint8 ** vector, cmph_uint32 nkeys)
{
	cmph_io_adapter_t * key_source = cmph_io_vector_new(vector, nkeys);
	key_source->read = key_byte_vector_read;
	key_source->dispose = key_vector_dispose;
	key_source->rewind = key_vector_rewind;
	return key_source;
}
void cmph_io_byte_vector_adapter_destroy(cmph_io_adapter_t * key_source)
{
	cmph_io_vector_destroy(key_source);
}

cmph_io_adapter_t *cmph_io_struct_vector_adapter(void * vector, cmph_uint32 struct_size, cmph_uint32 key_offset, cmph_uint32 key_len, cmph_uint32 nkeys)
{
	cmph_io_adapter_t * key_source = cmph_io_struct_vector_new(vector, struct_size, key_offset, key_len, nkeys);
	key_source->read = key_struct_vector_read;
	key_source->dispose = key_vector_dispose;
	key_source->rewind = key_struct_vector_rewind;
	return key_source;
}

void cmph_io_struct_vector_adapter_destroy(cmph_io_adapter_t * key_source)
{
	cmph_io_struct_vector_destroy(key_source);
}

cmph_io_adapter_t *cmph_io_vector_adapter(char ** vector, cmph_uint32 nkeys)
{
	cmph_io_adapter_t * key_source = cmph_io_vector_new(vector, nkeys);
	key_source->read = key_vector_read;
	key_source->dispose = key_vector_dispose;
	key_source->rewind = key_vector_rewind;
	return key_source;
}

void cmph_io_vector_adapter_destroy(cmph_io_adapter_t * key_source)
{
	cmph_io_vector_destroy(key_source);
}

cmph_config_t *cmph_config_new(cmph_io_adapter_t *key_source)
{
	cmph_config_t *mph = NULL;
	mph = __config_new(key_source);
	assert(mph);
	mph->algo = CMPH_CHM; // default value
	mph->data = chm_config_new();
	return mph;
}

void cmph_config_set_algo(cmph_config_t *mph, CMPH_ALGO algo)
{
	if (algo != mph->algo)
	{
		switch (mph->algo)
		{
			case CMPH_CHM:
				chm_config_destroy(mph);
				break;
			case CMPH_BMZ:
				bmz_config_destroy(mph);
				break;
			case CMPH_BMZ8:
				bmz8_config_destroy(mph);
				break;
			case CMPH_BRZ:
				brz_config_destroy(mph);
				break;
			case CMPH_FCH:
				fch_config_destroy(mph);
				break;
			case CMPH_BDZ:
				bdz_config_destroy(mph);
				break;
			case CMPH_BDZ_PH:
				bdz_ph_config_destroy(mph);
				break;
			case CMPH_CHD_PH:
				chd_ph_config_destroy(mph);
				break;
			case CMPH_CHD:
				chd_config_destroy(mph);
				break;
			default:
				assert(0);
		}
		switch(algo)
		{
			case CMPH_CHM:
				mph->data = chm_config_new();
				break;
			case CMPH_BMZ:
				mph->data = bmz_config_new();
				break;
			case CMPH_BMZ8:
				mph->data = bmz8_config_new();
				mph->nhashfuncs = 2;
				break;
			case CMPH_BRZ:
				mph->data = brz_config_new();
				break;
			case CMPH_FCH:
				mph->data = fch_config_new();
				break;
			case CMPH_BDZ:
				mph->data = bdz_config_new();
				break;
			case CMPH_BDZ_PH:
				mph->data = bdz_ph_config_new();
				break;
			case CMPH_CHD_PH:
				mph->data = chd_ph_config_new();
				break;
			case CMPH_CHD:
				mph->data = chd_config_new(mph);
				break;
			default:
				assert(0);
		}
	}
	mph->algo = algo;
}

void cmph_config_set_tmp_dir(cmph_config_t *mph, cmph_uint8 *tmp_dir)
{
	if (mph->algo == CMPH_BRZ)
	{
		brz_config_set_tmp_dir(mph, tmp_dir);
	}
}


void cmph_config_set_mphf_fd(cmph_config_t *mph, FILE *mphf_fd)
{
	if (mph->algo == CMPH_BRZ)
	{
		brz_config_set_mphf_fd(mph, mphf_fd);
	}
}

void cmph_config_set_b(cmph_config_t *mph, cmph_uint32 b)
{
	if (mph->algo == CMPH_BRZ)
	{
		brz_config_set_b(mph, b);
	}
	else if (mph->algo == CMPH_BDZ)
	{
		bdz_config_set_b(mph, b);
	}
	else if (mph->algo == CMPH_CHD_PH)
	{
		chd_ph_config_set_b(mph, b);
	}
	else if (mph->algo == CMPH_CHD)
	{
		chd_config_set_b(mph, b);
	}
}

void cmph_config_set_keys_per_bin(cmph_config_t *mph, cmph_uint32 keys_per_bin)
{
	if (mph->algo == CMPH_CHD_PH)
	{
		chd_ph_config_set_keys_per_bin(mph, keys_per_bin);
	}
	else if (mph->algo == CMPH_CHD)
	{
		chd_config_set_keys_per_bin(mph, keys_per_bin);
	}
}

void cmph_config_set_memory_availability(cmph_config_t *mph, cmph_uint32 memory_availability)
{
	if (mph->algo == CMPH_BRZ)
	{
		brz_config_set_memory_availability(mph, memory_availability);
	}
}

void cmph_config_destroy(cmph_config_t *mph)
{
	if(mph)
	{
		DEBUGP("Destroying mph with algo %s\n", cmph_names[mph->algo]);
		switch (mph->algo)
		{
			case CMPH_CHM:
				chm_config_destroy(mph);
				break;
			case CMPH_BMZ:
				bmz_config_destroy(mph);
				break;
			case CMPH_BMZ8:
				bmz8_config_destroy(mph);
				break;
			case CMPH_BRZ:
				brz_config_destroy(mph);
				break;
			case CMPH_FCH:
				fch_config_destroy(mph);
				break;
			case CMPH_BDZ:
				bdz_config_destroy(mph);
				break;
			case CMPH_BDZ_PH:
				bdz_ph_config_destroy(mph);
				break;
			case CMPH_CHD_PH:
				chd_ph_config_destroy(mph);
				break;
			case CMPH_CHD:
				chd_config_destroy(mph);
				break;
			default:
				assert(0);
		}
		__config_destroy(mph);
	}
}

void cmph_config_set_seed(cmph_config_t *mph, cmph_uint32 seed)
{
	mph->seed = seed;
}
void cmph_config_set_verbosity(cmph_config_t *mph, cmph_uint32 verbosity)
{
	mph->verbosity = verbosity & 0xFF;
}
void cmph_config_set_ordering_table(cmph_config_t *mph)
{
	mph->do_ordering_table = 1;
}

void cmph_config_set_hashfuncs(cmph_config_t *mph, CMPH_HASH *hashfuncs)
{
	switch (mph->algo)
	{
		case CMPH_CHM:
			chm_config_set_hashfuncs(mph, hashfuncs);
			break;
		case CMPH_BMZ:
			bmz_config_set_hashfuncs(mph, hashfuncs);
			break;
		case CMPH_BMZ8:
			bmz8_config_set_hashfuncs(mph, hashfuncs);
			break;
		case CMPH_BRZ:
			brz_config_set_hashfuncs(mph, hashfuncs);
			break;
		case CMPH_FCH:
			fch_config_set_hashfuncs(mph, hashfuncs);
			break;
		case CMPH_BDZ:
			bdz_config_set_hashfuncs(mph, hashfuncs);
			break;
		case CMPH_BDZ_PH:
			bdz_ph_config_set_hashfuncs(mph, hashfuncs);
			break;
		case CMPH_CHD_PH:
			chd_ph_config_set_hashfuncs(mph, hashfuncs);
			break;
		case CMPH_CHD:
			chd_config_set_hashfuncs(mph, hashfuncs);
			break;
		default:
			break;
	}
	return;
}
CMPH_HASH* cmph_config_hashfuncs(cmph_config_t *mph, cmph_uint32 *nhashfuncs)
{
    *nhashfuncs = mph->nhashfuncs;
    return mph->hashfuncs;
}

void cmph_config_set_graphsize(cmph_config_t *mph, double c)
{
	mph->c = c;
	return;
}

void cmph_config_set_c_prefix(cmph_config_t *mph, char *c_prefix)
{
	mph->c_prefix = c_prefix;
	return;
}

int cmph_is_order_preserving(cmph_t *mphf)
{
	switch (mphf->algo)
	{
		case CMPH_CHM:
		    return 1;
		case CMPH_BMZ:
		    return 0;
		case CMPH_BMZ8:
		    return 0;
		case CMPH_BRZ:
		    return 0;
		case CMPH_FCH:
		    return 0;
		case CMPH_BDZ:
		    return 0;
		case CMPH_BDZ_PH:
		    return 0;
		case CMPH_CHD_PH:
		    return 0;
		case CMPH_CHD:
		    return 0;
		default:
		    assert(0);
	}
	assert(0);
	return 0;
}

cmph_t *cmph_new(cmph_config_t *mph)
{
	cmph_t *mphf = NULL;
	double c = mph->c;

	DEBUGP("Creating mph with algorithm %s\n", cmph_names[mph->algo]);
	switch (mph->algo)
	{
		case CMPH_CHM:
			DEBUGP("Creating chm hash\n");
			mphf = chm_new(mph, c);
			break;
		case CMPH_BMZ:
			DEBUGP("Creating bmz hash\n");
			mphf = bmz_new(mph, c);
			break;
		case CMPH_BMZ8:
			DEBUGP("Creating bmz8 hash\n");
			mphf = bmz8_new(mph, c);
			break;
		case CMPH_BRZ:
			DEBUGP("Creating brz hash\n");
			if (c >= 2.0)
			    brz_config_set_algo(mph, CMPH_FCH);
			else
			    brz_config_set_algo(mph, CMPH_BMZ8);
			mphf = brz_new(mph, c);
			break;
		case CMPH_FCH:
			DEBUGP("Creating fch hash\n");
			mphf = fch_new(mph, c);
			break;
		case CMPH_BDZ:
			DEBUGP("Creating bdz hash\n");
			mphf = bdz_new(mph, c);
			break;
		case CMPH_BDZ_PH:
			DEBUGP("Creating bdz_ph hash\n");
			mphf = bdz_ph_new(mph, c);
			break;
		case CMPH_CHD_PH:
			DEBUGP("Creating chd_ph hash\n");
			mphf = chd_ph_new(mph, c);
			break;
		case CMPH_CHD:
			DEBUGP("Creating chd hash\n");
			mphf = chd_new(mph, c);
			break;
		default:
			assert(0);
	}
	return mphf;
}

int cmph_dump(cmph_t *mphf, FILE *f)
{
	switch (mphf->algo)
	{
		case CMPH_CHM:
			return chm_dump(mphf, f);
		case CMPH_BMZ:
			return bmz_dump(mphf, f);
		case CMPH_BMZ8:
			return bmz8_dump(mphf, f);
		case CMPH_BRZ:
			return brz_dump(mphf, f);
		case CMPH_FCH:
			return fch_dump(mphf, f);
		case CMPH_BDZ:
			return bdz_dump(mphf, f);
		case CMPH_BDZ_PH:
			return bdz_ph_dump(mphf, f);
		case CMPH_CHD_PH:
			return chd_ph_dump(mphf, f);
		case CMPH_CHD:
			return chd_dump(mphf, f);
		default:
			assert(0);
	}
	assert(0);
	return 0;
}
int cmph_compile(cmph_t *mphf, cmph_config_t *config, const char *c_file,
		 const char *keys_file)
{
	FILE *out = c_file ? fopen(c_file, "w") : stdout;
	if (!out) {
	    perror("fopen");
	    fprintf(stderr, "ERROR: Can't write to %s\n", c_file ? c_file : "stdout");
	    exit(1);
	}
	DEBUGP("Compiling mphf with algorithm %s\n", cmph_names[mphf->algo]);
	fprintf(out, "/* ex: set ro ft=c: -*- mode: c; buffer-read-only: t -*- */\n");
	fprintf(out, "/* Created via cmph -C ");
	if (config->verbosity)
	    fprintf(out, "-v ");
	if (mphf->algo != CMPH_CHM)
	    fprintf(out, "-a %s ", cmph_names[mphf->algo]);
	fprintf(out, "-s %u ", config->seed);
	for (unsigned i=0; i<config->nhashfuncs; ++i)
	{
	    if (config->hashfuncs[i] == CMPH_HASH_COUNT)
		break;
	    if (i>0 || config->hashfuncs[i] != CMPH_HASH_JENKINS)
		fprintf(out, "-f %s ", cmph_hash_names[config->hashfuncs[i]]);
	}
	fprintf(out, "\"%s\" */\n", keys_file);
	fprintf(out, "/* Do not modify */\n\n");
	fprintf(out, "/* n: %u */\n", config->key_source->nkeys);
	fprintf(out, "/* c: %f */\n", config->c);
	if (config->do_ordering_table)
	    fprintf(out, "#include <assert.h>\n");
	switch (mphf->algo)
	{
		case CMPH_CHM:
		    return chm_compile(mphf, config, out);
		case CMPH_BMZ:
		    return bmz_compile(mphf, config, out);
		case CMPH_BDZ:
		    return bdz_compile(mphf, config, out);
		case CMPH_FCH:
		    return fch_compile(mphf, config, out);
		case CMPH_BDZ_PH:
		    return bdz_ph_compile(mphf, config, out);
		case CMPH_CHD_PH:
		    return chd_ph_compile(mphf, config, out);
		case CMPH_CHD:
		    return chd_compile(mphf, config, out);
#ifdef DEBUG
		case CMPH_BMZ8:
		    return bmz8_compile(mphf, config, out);
		case CMPH_BRZ:
		    return brz_compile(mphf, config, out);
#endif
		default:
		    if (mphf->algo < CMPH_COUNT)
			fprintf(stderr, "-C not yet supported with algorithm %s\n", cmph_names[mphf->algo]);
		    else
			fprintf(stderr, "Illegal algorithm\n");
		    exit(1);
	}
	if (c_file)
	    fclose(out);
	assert(0);
	return 0;
}

cmph_t *cmph_load(FILE *f)
{
	cmph_t *mphf = NULL;
	DEBUGP("Loading mphf generic parts\n");
	mphf =  __cmph_load(f);
	if (mphf == NULL) return NULL;
	DEBUGP("Loading mphf algorithm dependent parts\n");

	switch (mphf->algo)
	{
		case CMPH_CHM:
			chm_load(f, mphf);
			break;
		case CMPH_BMZ:
			DEBUGP("Loading bmz algorithm dependent parts\n");
			bmz_load(f, mphf);
			break;
		case CMPH_BMZ8:
			DEBUGP("Loading bmz8 algorithm dependent parts\n");
			bmz8_load(f, mphf);
			break;
		case CMPH_BRZ:
			DEBUGP("Loading brz algorithm dependent parts\n");
			brz_load(f, mphf);
			break;
		case CMPH_FCH:
			DEBUGP("Loading fch algorithm dependent parts\n");
			fch_load(f, mphf);
			break;
		case CMPH_BDZ:
			DEBUGP("Loading bdz algorithm dependent parts\n");
			bdz_load(f, mphf);
			break;
		case CMPH_BDZ_PH:
			DEBUGP("Loading bdz_ph algorithm dependent parts\n");
			bdz_ph_load(f, mphf);
			break;
		case CMPH_CHD_PH:
			DEBUGP("Loading chd_ph algorithm dependent parts\n");
			chd_ph_load(f, mphf);
			break;
		case CMPH_CHD:
			DEBUGP("Loading chd algorithm dependent parts\n");
			chd_load(f, mphf);
			break;
		default:
			assert(0);
	}
	DEBUGP("Loaded mphf\n");
	if (mphf->size == 0) {
	    free(mphf);
	    return NULL;
	}
	return mphf;
}

cmph_uint32 *cmph_ordering_table(cmph_t *mphf)
{
	return mphf->o;
}

cmph_uint32 cmph_search(cmph_t *mphf, const char *key, cmph_uint32 keylen)
{
	//DEBUGP("mphf algorithm: %s (%u)\n", cmph_names[mphf->algo], mphf->algo);
	switch(mphf->algo)
	{
		case CMPH_CHM:
			return chm_search(mphf, key, keylen);
		case CMPH_BMZ:
			//DEBUGP("bmz algorithm search\n");
		        return bmz_search(mphf, key, keylen);
		case CMPH_BMZ8:
			//DEBUGP("bmz8 algorithm search\n");
		        return bmz8_search(mphf, key, keylen);
		case CMPH_BRZ:
			//DEBUGP("brz algorithm search\n");
		        return brz_search(mphf, key, keylen);
		case CMPH_FCH:
		    	//DEBUGP("fch algorithm search\n");
		        return fch_search(mphf, key, keylen);
		case CMPH_BDZ:
		        //DEBUGP("bdz algorithm search\n");
		        return bdz_search(mphf, key, keylen);
		case CMPH_BDZ_PH:
			//DEBUGP("bdz_ph algorithm search\n");
		        return bdz_ph_search(mphf, key, keylen);
		case CMPH_CHD_PH:
			//DEBUGP("chd_ph algorithm search\n");
		        return chd_ph_search(mphf, key, keylen);
		case CMPH_CHD:
			//DEBUGP("chd algorithm search\n");
		        return chd_search(mphf, key, keylen);
		default:
			assert(0);
	}
	assert(0);
	return 0;
}

cmph_uint32 cmph_size(cmph_t *mphf)
{
	return mphf->size;
}

void cmph_destroy(cmph_t *mphf)
{
	switch(mphf->algo)
	{
		case CMPH_CHM:
			chm_destroy(mphf);
			return;
		case CMPH_BMZ:
			bmz_destroy(mphf);
			return;
		case CMPH_BMZ8:
			bmz8_destroy(mphf);
			return;
		case CMPH_BRZ:
			brz_destroy(mphf);
			return;
		case CMPH_FCH:
			fch_destroy(mphf);
			return;
		case CMPH_BDZ:
			bdz_destroy(mphf);
			return;
		case CMPH_BDZ_PH:
			bdz_ph_destroy(mphf);
			return;
		case CMPH_CHD_PH:
			chd_ph_destroy(mphf);
			return;
		case CMPH_CHD:
			chd_destroy(mphf);
			return;
		default:
			assert(0);
	}
	assert(0);
	return;
}


/** \fn void cmph_pack(cmph_t *mphf, void *packed_mphf);
 *  \brief Support the ability to pack a perfect hash function into a preallocated contiguous memory space pointed by packed_mphf.
 *  \param mphf pointer to the resulting mphf
 *  \param packed_mphf pointer to the contiguous memory area used to store the resulting mphf. The size of packed_mphf must be at least cmph_packed_size()
 */
void cmph_pack(cmph_t *mphf, void *packed_mphf)
{
	// packing algorithm type to be used in cmph.c
	cmph_uint32 * ptr = (cmph_uint32 *) packed_mphf;
	*ptr++ = mphf->algo;
	DEBUGP("mphf->algo = %u\n", mphf->algo);
	switch(mphf->algo)
	{
		case CMPH_CHM:
			chm_pack(mphf, ptr);
			break;
		case CMPH_BMZ:
			bmz_pack(mphf, ptr);
			break;
		case CMPH_BMZ8:
			bmz8_pack(mphf, ptr);
			break;
		case CMPH_BRZ:
			brz_pack(mphf, ptr);
			break;
		case CMPH_FCH:
			fch_pack(mphf, ptr);
			break;
		case CMPH_BDZ:
			bdz_pack(mphf, ptr);
			break;
		case CMPH_BDZ_PH:
			bdz_ph_pack(mphf, ptr);
			break;
		case CMPH_CHD_PH:
			chd_ph_pack(mphf, ptr);
			break;
		case CMPH_CHD:
			chd_pack(mphf, ptr);
			break;
		default:
			assert(0);
	}
	return;
}

/** \fn cmph_uint32 cmph_packed_size(cmph_t *mphf);
 *  \brief Return the amount of space needed to pack mphf.
 *  \param mphf pointer to a mphf
 *  \return the size of the packed function or zero for failures
 */
cmph_uint32 cmph_packed_size(cmph_t *mphf)
{
	switch(mphf->algo)
	{
		case CMPH_CHM:
			return chm_packed_size(mphf);
		case CMPH_BMZ:
			return bmz_packed_size(mphf);
		case CMPH_BMZ8:
			return bmz8_packed_size(mphf);
		case CMPH_BRZ:
			return brz_packed_size(mphf);
		case CMPH_FCH:
			return fch_packed_size(mphf);
		case CMPH_BDZ:
			return bdz_packed_size(mphf);
		case CMPH_BDZ_PH:
			return bdz_ph_packed_size(mphf);
		case CMPH_CHD_PH:
			return chd_ph_packed_size(mphf);
		case CMPH_CHD:
			return chd_packed_size(mphf);
		default:
			assert(0);
	}
	return 0; // FAILURE
}

/** cmph_uint32 cmph_search(void *packed_mphf, const char *key, cmph_uint32 keylen);
 *  \brief Use the packed mphf to do a search.
 *  \param  packed_mphf pointer to the packed mphf
 *  \param key key to be hashed
 *  \param keylen key legth in bytes
 *  \return The mphf value
 */
cmph_uint32 cmph_search_packed(void *packed_mphf, const char *key, cmph_uint32 keylen)
{
	cmph_uint32 *ptr = (cmph_uint32 *)packed_mphf;
//	fprintf(stderr, "algo:%u\n", *ptr);
	switch(*ptr)
	{
		case CMPH_CHM:
			return chm_search_packed(++ptr, key, keylen);
		case CMPH_BMZ:
			return bmz_search_packed(++ptr, key, keylen);
		case CMPH_BMZ8:
			return bmz8_search_packed(++ptr, key, keylen);
		case CMPH_BRZ:
			return brz_search_packed(++ptr, key, keylen);
		case CMPH_FCH:
			return fch_search_packed(++ptr, key, keylen);
		case CMPH_BDZ:
			return bdz_search_packed(++ptr, key, keylen);
		case CMPH_BDZ_PH:
			return bdz_ph_search_packed(++ptr, key, keylen);
		case CMPH_CHD_PH:
			return chd_ph_search_packed(++ptr, key, keylen);
		case CMPH_CHD:
			return chd_search_packed(++ptr, key, keylen);
		default:
			assert(0);
	}
	return 0; // FAILURE
}
